model/fn-execution/src/main/proto/beam_fn_api.proto

```diff
@@ -501,48 +501,9 @@ message MonitoringInfoTypeUrns {
 message Metric {
   // (Required) The data for this metric.
   oneof data {
-    CounterData counter_data = 1;
-    DistributionData distribution_data = 2;
-    ExtremaData extrema_data = 3;
-  }
-}
-
-// Data associated with a Counter or Gauge metric.
-// This is designed to be compatible with metric collection
-// systems such as DropWizard.
-message CounterData {
-  oneof value {
-    int64 int64_value = 1;
-    double double_value = 2;
-    string string_value = 3;
-  }
-}
-
-// Extrema messages are used for calculating
-// Top-N/Bottom-N metrics.
-message ExtremaData {
-  oneof extrema {
-    IntExtremaData int_extrema_data = 1;
-    DoubleExtremaData double_extrema_data = 2;
-  }
-}
-
-message IntExtremaData {
-  repeated int64 int_values = 1;
-}
-
-message DoubleExtremaData {
-  repeated double double_values = 2;
-}
-
-// Data associated with a distribution metric.
-// This is based off of the current DistributionData metric.
-// This is not a stackdriver or dropwizard compatible
-// style of distribution metric.
-message DistributionData {
-  oneof distribution {
-    IntDistributionData int_distribution_data = 1;
-    DoubleDistributionData double_distribution_data = 2;
+    int64 counter = 1;
```

Comment:
There was long debate when, and the existing protos were what we could agree on when we introduced this design. While I am fine with this change you are proposing personally. I'd like you to get buy in on the dev list.

https://s.apache.org/beam-fn-api-metrics

https://lists.apache.org/list.html?dev@beam.apache.org:2018-2

I think most of it was in the comments of the doc. So unfortunately, I cannot point you toward too much.

Reply:
I don't have strong opinion about removing CounterData, but I'm concerned that we make counter int64. In my opinion, double will be much more flexible.

Reply:
I don't remember discussing this, it must have been with someone else. 

I agree that supporting double counters (and especially distributions) would be useful. Also, we haven't really nailed down what the global semantics are for gauge metrics. 

My position was, and remains, that reifying the exact set of supported metric types via a set of nested set of protos is not a stable, long-term solution. It took a long time to come up with this list, and here we are 6 months later changing it. (And change will become harder the more we actually use these protos everywhere.) Instead, data should be either (1) a bytes field or (2) something like

message data {
  repeated int64 int_data = 1;
  repeated double double_data = 2;
  repeated string string_data = 3;
  ...?
}

with the type specifying its encoding into this data attribute. We can then support everything from simple counters to complex distributions without changing the protos, only introducing new types, where the set of types are defined by URNs, and runners simply pass through types they do not understand. This is consistent with the rest of the way the FnAPI works as well (e.g. coders, transforms, ...)

Reply:
Thanks Robert. It seems we're stuck between 3 data models:

1. [existing protos](https://github.com/apache/beam/blob/981fb620a1b53d2a564b629f1d1e2d18184f8393/model/fn-execution/src/main/proto/beam_fn_api.proto#L501-L561)
    - {int,double,string} counters
    - {int,double} x {distribution,extrema}
    - unused in SDKs
2. [deprecated protos](https://github.com/apache/beam/blob/981fb620a1b53d2a564b629f1d1e2d18184f8393/model/fn-execution/src/main/proto/beam_fn_api.proto#L687-L714)
    - `int64`-only {counter, distribution, gauge}
    - supported in [Java](https://github.com/apache/beam/blob/981fb620a1b53d2a564b629f1d1e2d18184f8393/sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricQueryResults.java#L32-L39), [Python](https://github.com/apache/beam/blob/981fb620a1b53d2a564b629f1d1e2d18184f8393/sdks/python/apache_beam/metrics/cells.py) SDKs
3. opaque bytes-payloads
    - [Robert's suggestion above](https://github.com/apache/beam/pull/7876#discussion_r258384748)
    - not implemented anywhere afaik

In this PR, I changed (1)'s protos to match the SDK support that is modeled on (2), which is probably wrong!

(3) sgtm but seems like it is not the existing consensus / spec.

I think the larger goal (sending metrics over the job API) requires converting between SDK and proto metrics structures (in both directions), so they need to match.

Maybe there is a way I can side-step that, or maybe we can just get to the real long-term solution here.

Reply:
@robertwb I'd like to push back a bit about a significant change, where we use opaque bytes payloads.

I am okay with removing the layers in this proto and using the design Ryan has proposed here. But not an extensive rewrite at this stage of the game, which will slow down progress significantly. 

We have protos for the MonitoringInfoTable. This solution is the extensible format, where you can store basically anything and the producer and consumer of this proto can be the only ones who need to understand it (and it could be passed through the RunnerHarness)
https://github.com/apache/beam/blob/197a06852d27c5baea6a4c65894a8841672af7b6/model/fn-execution/src/main/proto/beam_fn_api.proto#L583

But we chose to have separate protos for the common well known metric formats (Metrics being a well defined concept shared by many systems, i.e. a timeseries of data): counter, gauge, distribution, etc.

Additionally, if you wish to modify these formats significantly at a future stage, you can always 
- add to these protos without deleting
- Change the URN version number of a metric
- Upgrade all SDKs to package the metric with a format for the previous and new version number of the URN
- Upgrade all Runners that wish to support it.

Upgrading the URN version is the safe way to do this without introducing a breaking change, as consumers of MonitoringInfos can freely chose which URNs+versions they support.